src/cpp/core/file_lock/LinkBasedFileLock.cpp

```diff
@@ -0,0 +1,318 @@
+         pidString()     + "-" +
+         threadId();
+         
+}
+
+struct CloseFileScope
+{
+   CloseFileScope(int fd, ErrorLocation location)
+      : fd_(fd), location_(location)
+   {}
+   
+   ~CloseFileScope()
+   {
+      int errc = ::close(fd_);
+      if (errc)
+         LOG_ERROR(systemError(errno, location_));
+   }
+   
+   int fd_;
+   ErrorLocation location_;
+};
+
+class LockRegistration : boost::noncopyable
+{
+public:
+   
+   void registerLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.insert(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void deregisterLock(const FilePath& lockFilePath)
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         registration_.erase(lockFilePath);
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void refreshLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            lockFilePath.setLastWriteTime();
+         }
+      }
+      END_LOCK_MUTEX
+   }
+   
+   void clearLocks()
+   {
+      LOCK_MUTEX(mutex_)
+      {
+         BOOST_FOREACH(const FilePath& lockFilePath, registration_)
+         {
+            Error error = lockFilePath.removeIfExists();
+            if (error)
+               LOG_ERROR(error);
+         }
+         registration_.clear();
+      }
+      END_LOCK_MUTEX
+   }
+   
+   ~LockRegistration()
+   {
+      try
+      {
+         clearLocks();
+      }
+      catch (...)
+      {
+      }
+   }
+   
+private:
+   
+   boost::mutex mutex_;
+   std::set<FilePath> registration_;
+   
+};
+
+LockRegistration& lockRegistration()
+{
+   static LockRegistration instance;
+   return instance;
+}
+
+Error writeLockFile(const FilePath& lockFilePath)
+{
+   
+#ifndef _WIN32
+   
+   // generate proxy lockfile
+   FilePath proxyPath = lockFilePath.parent().complete(proxyLockFileName());
+   if (proxyPath.exists())
+      return fileExistsError(ERROR_LOCATION);
+   
+   Error error = core::writeStringToFile(proxyPath, "");
+   if (error)
+      return error;
+   RemoveOnExitScope scope(proxyPath, ERROR_LOCATION);
+   
+   // attempt to link to the desired location -- ignore return value
+   // and just stat our original link after
+   ::link(
+            proxyPath.absolutePathNative().c_str(),
+            lockFilePath.absolutePathNative().c_str());
+   
+   struct stat info;
+   int errc = ::stat(proxyPath.absolutePathNative().c_str(), &info);
+   if (errc)
+      return systemError(errno, ERROR_LOCATION);
+   
+   if (info.st_nlink != 2)
+      return fileExistsError(ERROR_LOCATION);
+   
+   return Success();
+   
+#else
+   
+   return -1;
+   
+#endif
+}
+
+bool isLockFileStale(const FilePath& lockFilePath, double seconds = 30.0)
+{
+   double diff = ::difftime(::time(NULL), lockFilePath.lastWriteTime());
+   return diff >= seconds;
+}
+
+Error beginLocking(const FilePath& lockingPath)
+{
+   // attempt to clean up stale lockfiles
```

Comment:
Do we have a mechanism for cleaning up abandoned lock files? (i.e. lock files from other processes that were abandoned due to a crash or restart). It seems like somewhere we should have code that looks at all the adjacent locks files and removes any that are stale.